Fixes for Windows

cwenger opened the pull request:
Main changes:
- `long` is only 4 bytes on Windows (32- and 64-bit; [source](https://software.intel.com/content/www/us/en/develop/articles/size-of-long-integer-type-on-different-architecture-and-os.html)), so switched to `long long` which should always be 8 bytes
- explicitly open .hic file as binary to prevent issues with newline differences

Also a few minor fixes for compiling with Microsoft Visual Studio.

nchernia reviewed (commented):
I haven't tested on Unix, @sa501428  could you see?

I think you did a search and replace, line 88 is straw-R.cpp now has a typo.

I am wondering if we should move to stdint.h as suggested here: https://stackoverflow.com/questions/6462439/whats-the-difference-between-long-long-and-long

sa501428 commented:
I do prefer the idea of defining the lengths, especially as they are technically decided by the hic file format, not the length of particular variable types.

Other major issue is that we actually have a pretty substantial restructuring of straw to better handle more advanced use cases: https://github.com/sa501428/straw Can you actually make the pull request to that repo. After we finish up some testing, that fork will get merged here and overwrite the master branch here.

cwenger commented:
> I think you did a search and replace, line 88 is straw-R.cpp now has a typo.

Yes good catch, I fixed that in the native C++ code but not the R C++ code! Corrected now.

sa501428 commented:
@cwenger please see the changes pulled into the master branch. There is a lot of restructuring internally. Old usage is still supported, but new usage has also been added.

cwenger commented:
Thanks @sa501428, will do!

sa501428 commented:
Just posting this reference here for the definitions used in the hic file format

Strings are null (0) terminated. So for example the string "HIC" is represented by 4 bytes [48 49 43 0]
short - 16 bit integer
int - 32 bit integer
long - 64 bit integer
float - 32 bit floating point
double - 64 bit floating point

sa501428 commented on C++/straw.cpp:
```diff
@@ -760,18 +760,18 @@ vector<contactRecord> readBlock(istream &fin, CURL *curl, bool isHttp, indexEntr
     }
 
-    unsigned long numValues = static_cast<unsigned long>(nValues);
+    unsigned long long numValues = static_cast<unsigned long>(nValues);
```
should it be `static_cast<unsigned long long>`?

cwenger replied:
Fixed; thank you.

sa501428 commented on C++/straw.cpp:
```diff
@@ -927,12 +927,12 @@ class HiCFile {
 };
 
-long HiCFile::totalFileSize = 0L;
+long long HiCFile::totalFileSize = 0L;
```
should this be 0LL (and default in other places as well?)

cwenger replied:
Fixed; thank you.

sa501428 commented on R/DESCRIPTION:
```diff
@@ -1,6 +1,6 @@
 Package: strawr
 Title: Straw is a fast implementation of reading/dump for .hic files
-Version: 0.0.3
+Version: 0.0.4
```
thoughts on bumping this to 0.0.8 to match pybind11 version?

cwenger replied:
Good idea; done.

sa501428 commented on C++/main.cpp:
```diff
@@ -23,6 +23,7 @@
 */
 #include <cstring>
 #include <iostream>
+#include <string>
```
if we do this can we get rid of the `include <cstring>` above?

cwenger replied:
Yes that seems to be true; done.

sa501428 reviewed (commented):
Some questions/thoughts

sa501428 commented:
Actually I'm confused now - how does this function still work?
```
long long readLongLongFromFile(istream &fin) {
    long long tempLong;
    fin.read((char *) &tempLong, sizeof(long));
    return tempLong;
}
```
Given it still ends up calling sizeof(long) and not sizeof(long long)?

cwenger commented:
@sa501428 You are correct. I missed a couple longs when I was re-implementing the changes, and I only tested in Windows with R, not with the command-line version. I'll do another PR to fix this.
